File: docs/app/extract-signature.js
/**
 * The signature extractor page.
 *
 * The command-line version takes the crop as four fractions you read off a grid
 * image. On a page you can just drag a box over the photograph, so that is what this
 * does; the numbers underneath are the same ones extract_signature.py takes.
 */

import { extractSignature } from "../imaging.js";
import {
  $, dropZone, status, yieldToPaint, loadImage, pixelsOf, pixelsFromPdf, isPdf,
  canvasOf, blobOf, offerDownload, humanSize,
} from "./ui.js";

const state = $("state");
const stage = $("stage");
const shot = $("shot");
const box = $("box");
const result = $("result");
const save = $("save");

// Where a signature sits on a signed document: right half, lower band. Same default
// as the CLI, expressed as fractions so it holds at any resolution.
let region = { x0: 0.45, y0: 0.70, x1: 1.00, y1: 0.97 };
let source = null;
let name = "signature";

dropZone($("drop"), $("picker"), files => open(files[0]));
for (const id of ["light", "dark", "colour"]) {
  $(id).addEventListener("input", () => { if (source) extract(); });
}

async function open(file) {
  try {
    status(state, "Reading the photograph…");
    await yieldToPaint();
    name = file.name.replace(/\.[^.]+$/, "");
    source = isPdf(file) ? await pixelsFromPdf(file, 2400)
                         : pixelsOf(await loadImage(file), 2400);
    const canvas = canvasOf(source.data, source.width, source.height);
    shot.width = source.width;
    shot.height = source.height;
    shot.getContext("2d").drawImage(canvas, 0, 0);
    stage.hidden = false;
    save.hidden = true;
    result.innerHTML = "";
    drawBox();
    // No extraction yet. The default box is a guess about where a signature usually
    // sits, and guessing wrong greets you with a red warning before you have done
    // anything. Ask for the box first.
    status(state, "Now drag a box around the signature.");
  } catch (err) {
    status(state, `Could not read that image: ${err && err.message || err}`, "bad");
  }
}

// --- the draggable box ------------------------------------------------------

function drawBox() {
  box.style.left = `${region.x0 * 100}%`;
  box.style.top = `${region.y0 * 100}%`;
  box.style.width = `${(region.x1 - region.x0) * 100}%`;
  box.style.height = `${(region.y1 - region.y0) * 100}%`;
  $("fractions").textContent =
    `--box ${region.x0.toFixed(2)},${region.y0.toFixed(2)},${region.x1.toFixed(2)},${region.y1.toFixed(2)}`;
}

const clamp = v => Math.min(1, Math.max(0, v));

function pointIn(event) {
  const r = stage.getBoundingClientRect();
  return { x: clamp((event.clientX - r.left) / r.width), y: clamp((event.clientY - r.top) / r.height) };
}

let drag = null;

stage.addEventListener("pointerdown", e => {
  if (!source) return;
  const p = pointIn(e);
  const onHandle = e.target === $("handle");
  const inside = !onHandle && e.target === box;
  drag = { mode: onHandle ? "resize" : inside ? "move" : "draw", from: p, start: { ...region } };
  if (drag.mode === "draw") region = { x0: p.x, y0: p.y, x1: p.x, y1: p.y };
  stage.setPointerCapture(e.pointerId);
  e.preventDefault();
});

stage.addEventListener("pointermove", e => {
  if (!drag) return;
  const p = pointIn(e);
  if (drag.mode === "draw") {
    region = { x0: Math.min(drag.from.x, p.x), y0: Math.min(drag.from.y, p.y),
               x1: Math.max(drag.from.x, p.x), y1: Math.max(drag.from.y, p.y) };
  } else if (drag.mode === "move") {
    const dx = p.x - drag.from.x, dy = p.y - drag.from.y;
    const w = drag.start.x1 - drag.start.x0, h = drag.start.y1 - drag.start.y0;
    const x0 = Math.min(Math.max(drag.start.x0 + dx, 0), 1 - w);
    const y0 = Math.min(Math.max(drag.start.y0 + dy, 0), 1 - h);
    region = { x0, y0, x1: x0 + w, y1: y0 + h };
  } else {
    region = { ...region, x1: Math.max(p.x, region.x0 + 0.02), y1: Math.max(p.y, region.y0 + 0.02) };
  }
  drawBox();
});

stage.addEventListener("pointerup", async () => {
  if (!drag) return;
  drag = null;
  if (region.x1 - region.x0 > 0.02 && region.y1 - region.y0 > 0.02) await extract();
});

// --- the extraction ---------------------------------------------------------

async function extract() {
  status(state, "Cutting it out…");
  await yieldToPaint();

  const x0 = Math.floor(region.x0 * source.width), y0 = Math.floor(region.y0 * source.height);
  const x1 = Math.max(x0 + 2, Math.floor(region.x1 * source.width));
  const y1 = Math.max(y0 + 2, Math.floor(region.y1 * source.height));
  const w = x1 - x0, h = y1 - y0;

  const crop = new Uint8ClampedArray(w * h * 4);
  for (let y = 0; y < h; y++) {
    const from = ((y0 + y) * source.width + x0) * 4;
    crop.set(source.data.subarray(from, from + w * 4), y * w * 4);
  }

  const cut = extractSignature(crop, w, h, {
    light: Number($("light").value),
    dark: Number($("dark").value),
    keepColour: $("colour").checked,
  });

  const full = canvasOf(cut.rgba, w, h);
  const b = cut.box || { x0: 0, y0: 0, x1: w, y1: h };
  const trimmed = document.createElement("canvas");
  trimmed.width = b.x1 - b.x0;
  trimmed.height = b.y1 - b.y0;
  trimmed.getContext("2d").drawImage(full, -b.x0, -b.y0);

  const blob = await blobOf(trimmed, "image/png");
  result.innerHTML = "";
  result.appendChild(trimmed);
  offerDownload(save, blob, `${name}_signature.png`);
  save.textContent = `Save the PNG (${humanSize(blob.size)})`;
  save.hidden = false;

  const coverage = cut.coverage;
  if (coverage < 0.4) {
    status(state, `Almost no ink in that box (${coverage.toFixed(1)}%). The signature is ` +
                  "somewhere else on the photo. Drag the box over it.", "bad");
  } else if (coverage > 25) {
    status(state, `That is a lot of ink (${coverage.toFixed(1)}%): you have caught body ` +
                  "text or the shadow of the sheet. Tighten the box, or lower the paper " +
                  "threshold.", "bad");
  } else {
    status(state, `Ink in the box: ${coverage.toFixed(1)}%. Check the preview against the ` +
                  "chequerboard: only the signature should be showing.", "good");
  }
  drawBox();
}

File: tools/extract_signature.py
#!/usr/bin/env python3
"""
Cut a handwritten signature out of a photograph and save it as a transparent PNG.

What it is for: it is the missing piece for having a real PDF instead of a
photograph of one. Shooting the whole sheet gives you an image — soft text, heavy
file, no selectable text. Taking only the signature and pasting it onto the original
digital PDF keeps the document crisp and machine-readable, while the signature is
still genuine ink written by hand.

How the strokes are separated from the background: the illumination is corrected
first (the same technique as clean_scan.py), then the ALPHA channel is built from
how dark each pixel is. Light paper becomes transparent, dark ink becomes opaque,
and the values in between stay semi-transparent, which keeps the edges smooth. A
hard threshold would produce a jagged signature that looks obviously cut out.

Usage:
    python3 extract_signature.py signed.pdf -o signature.png
    python3 extract_signature.py photo.jpg --box 0.5,0.75,1.0,0.95 -o signature.png
    python3 extract_signature.py photo.jpg --blue -o signature.png   # keep blue ink
    python3 extract_signature.py photo.jpg --preview grid.png        # pick the box

Then paste it onto the digital document:
    python3 add_date_signature.py document.pdf --signature signature.png
"""

import argparse
from pathlib import Path

import numpy as np
from PIL import Image, ImageDraw

from clean_scan import light_field, load

# Where a signature sits on a signed document: right half, lower band. Expressed as
# fractions of the page so it holds at any photo resolution.
DEFAULT_BOX = (0.45, 0.70, 1.00, 0.97)


def remove_long_runs(alpha: np.ndarray, fraction: float) -> np.ndarray:
    """Erase the printed rule under the signature, keeping the pen strokes.

    MEASURED, AND IT DOES NOT WORK on a skewed photo — which is why the default is
    0, meaning off. On the case this was built for, the rule was tilted just enough
    that its longest unbroken run inside any single pixel row reached 75 out of 508,
    about 15 per cent, while the signature's own strokes reached 66. The two
    populations overlap, so no threshold removes the rule without eating the
    signature. Separating them properly would mean deskewing and running a line
    detector (Hough), which is another order of complexity.

    It is useful only on a straight scan, where the rule crosses few rows and fills
    them. The clean path is different anyway: sign on a BLANK sheet, where there is
    no rule to separate in the first place.
    """
    limit = max(int(alpha.shape[1] * fraction), 8)
    dark = (alpha > 20).astype(np.int8)
    for y in range(dark.shape[0]):
        edges = np.flatnonzero(np.diff(np.concatenate(([0], dark[y], [0]))))
        for start, end in zip(edges[::2], edges[1::2]):
            if end - start >= limit:
                alpha[y, start:end] = 0
    return alpha


def extract(img: Image.Image, box, light: float, dark: float, blue: bool,
            run_fraction: float = 0.0):
    w, h = img.size
    x0, y0, x1, y1 = (int(box[0] * w), int(box[1] * h), int(box[2] * w), int(box[3] * h))
    crop = img.crop((x0, y0, x1, y1))

    grey = np.asarray(crop.convert("L"), dtype=np.float32)
    illumination = np.maximum(light_field(grey), 1.0)
    flat = np.clip(grey * (255.0 / illumination), 0, 255)

    # A ramp, not a step: above 'light' it is paper (transparent), below 'dark' it
    # is solid ink (opaque), in between it fades. These are the two numbers to touch
    # if the signature comes out faint (raise 'light') or if background grime
    # survives (lower it).
    alpha = np.clip((light - flat) / max(light - dark, 1.0) * 255.0, 0, 255)

    if run_fraction:
        alpha = remove_long_runs(alpha, run_fraction)

    if blue:
        rgb = np.asarray(crop.convert("RGB"), dtype=np.float32)
        colour = np.clip(rgb * (255.0 / illumination)[:, :, None] * 0.75, 0, 255).astype(np.uint8)
    else:
        colour = np.zeros(flat.shape + (3,), dtype=np.uint8)   # black ink

    out = Image.fromarray(np.dstack([colour, alpha.astype(np.uint8)]), mode="RGBA")

    # Trim to the content: a PNG carrying half a page of transparency around the
    # strokes would scale badly when laid on a signature line.
    bbox = Image.fromarray((alpha > 30).astype(np.uint8) * 255).getbbox()
    if bbox:
        pad = 6
        out = out.crop((max(bbox[0] - pad, 0), max(bbox[1] - pad, 0),
                        min(bbox[2] + pad, out.width), min(bbox[3] + pad, out.height)))
    return out, alpha


def write_grid(img: Image.Image, path: str):
    """Save a copy with a labelled grid, to read the fractions for --box off it.

    Faster than guessing: red lines are x, blue lines are y, every 0.05.
    """
    g = img.convert("RGB").copy()
    d = ImageDraw.Draw(g)
    for i in range(1, 20):
        f = i / 20
        d.line([(f * g.width, 0), (f * g.width, g.height)], fill=(255, 0, 0), width=1)
        d.line([(0, f * g.height), (g.width, f * g.height)], fill=(0, 120, 255), width=1)
        d.text((f * g.width + 2, 2), f"{f:.2f}", fill=(255, 0, 0))
        d.text((2, f * g.height + 2), f"{f:.2f}", fill=(0, 120, 255))
    g.save(path)


def main():
    p = argparse.ArgumentParser(
        description="Extract a signature from a photograph -> transparent PNG.")
    p.add_argument("input", help="photo or PDF of the signed document")
    p.add_argument("-o", "--output", default="signature.png")
    p.add_argument("--box", default=None,
                   help="area to look in, as fractions: x0,y0,x1,y1 (default: lower right)")
    p.add_argument("--light", type=float, default=205.0,
                   help="above this value it is paper; raise it if the signature comes out faint")
    p.add_argument("--dark", type=float, default=95.0,
                   help="below this value it is solid ink")
    p.add_argument("--blue", action="store_true", help="keep the pen colour")
    p.add_argument("--line-fraction", type=float, default=0.0,
                   help="erase horizontal runs longer than this fraction. Only works on a "
                        "straight scan; on a skewed photo it does not, see the code")
    p.add_argument("--preview", default=None,
                   help="save a grid image with the fractions, to choose --box")
    a = p.parse_args()

    src = Path(a.input)
    if not src.is_file():
        p.error(f"file not found: {src}")

    box = DEFAULT_BOX
    if a.box:
        box = tuple(float(v) for v in a.box.split(","))
        if len(box) != 4:
            p.error("--box takes four numbers: x0,y0,x1,y1")

    img = load(src)

    if a.preview:
        write_grid(img, a.preview)
        print(f"Grid saved to {a.preview}: red is x, blue is y.")
        print("Read off the corners of the signature and rerun with --box x0,y0,x1,y1")
        return

    signature, alpha = extract(img, box, a.light, a.dark, a.blue, a.line_fraction)
    signature.save(a.output)

    coverage = float((alpha > 30).mean()) * 100
    print(f"Wrote: {a.output}  ({signature.width}x{signature.height} px)")
    print(f"Ink found in the area: {coverage:.1f}%")
    if coverage < 0.4:
        print("  Almost nothing: the signature is not in that area. Pass --box with the "
              "right fractions, e.g. --box 0.4,0.6,1.0,0.95, or use --preview first.")
    elif coverage > 25:
        print("  Too much: you have caught body text or the shadow of the sheet. "
              "Tighten the box, or lower --light.")
    print("Open it and look before using it: only the signature should be visible, "
          "on a checkerboard (transparent) background.")


if __name__ == "__main__":
    main()
